File: test/contact_test.go
package test

import (
	"strings"
	"testing"
)

// --- Contact CRUD ---

func TestContactAdd_Basic(t *testing.T) {
	tc := newTestContext(t)

	// Adding a contact with just a name should succeed and print an ID.
	out := tc.runOK("contact", "add", "--name", "Jane Doe")
	out = strings.TrimSpace(out)
	if !strings.HasPrefix(out, "ct_") {
		t.Fatalf("expected contact ID starting with ct_, got: %s", out)
	}
}

func TestContactAdd_Full(t *testing.T) {
	tc := newTestContext(t)

	out := tc.runOK("contact", "add",
		"--name", "Jane Doe",
		"--email", "jane@acme.com",
		"--phone", "+1-555-0100",
		"--company", "Acme Corp",
		"--title", "CTO",
		"--source", "conference",
		"--tag", "hot-lead",
		"--tag", "enterprise",
		"--set", "linkedin=linkedin.com/in/janedoe",
	)
	id := strings.TrimSpace(out)
	if !strings.HasPrefix(id, "ct_") {
		t.Fatalf("expected contact ID, got: %s", id)
	}

	// Verify the contact was stored correctly.
	show := tc.runOK("contact", "show", id)
	assertContains(t, show, "Jane Doe")
	assertContains(t, show, "jane@acme.com")
	assertContains(t, show, "+1-555-0100")
	assertContains(t, show, "CTO")
	assertContains(t, show, "conference")
	assertContains(t, show, "hot-lead")
	assertContains(t, show, "enterprise")
	assertContains(t, show, "linkedin.com/in/janedoe")
}

func TestContactAdd_NameRequired(t *testing.T) {
	tc := newTestContext(t)

	// Should fail without --name.
	_, stderr := tc.runFail("contact", "add", "--email", "nobody@example.com")
	assertContains(t, stderr, "name")
}

func TestContactAdd_DuplicateEmail(t *testing.T) {
	tc := newTestContext(t)

	tc.runOK("contact", "add", "--name", "Jane Doe", "--email", "jane@acme.com")
	// Adding another contact with the same email should fail.
	_, stderr := tc.runFail("contact", "add", "--name", "Jane Smith", "--email", "jane@acme.com")
	assertContains(t, stderr, "duplicate")
}

func TestContactShow_ByEmail(t *testing.T) {
	tc := newTestContext(t)

	tc.runOK("contact", "add", "--name", "Jane Doe", "--email", "jane@acme.com")
	out := tc.runOK("contact", "show", "jane@acme.com")
	assertContains(t, out, "Jane Doe")
}

func TestContactShow_NotFound(t *testing.T) {
	tc := newTestContext(t)

	_, _ = tc.runFail("contact", "show", "nonexistent@example.com")
}

func TestContactList_Empty(t *testing.T) {
	tc := newTestContext(t)

	out := tc.runOK("contact", "list", "--format", "json")
	if strings.TrimSpace(out) != "[]" {
		t.Fatalf("expected empty JSON array, got: %s", out)
	}
}

func TestContactList_Multiple(t *testing.T) {
	tc := newTestContext(t)

	tc.runOK("contact", "add", "--name", "Alice", "--email", "alice@example.com")
	tc.runOK("contact", "add", "--name", "Bob", "--email", "bob@example.com")
	tc.runOK("contact", "add", "--name", "Charlie", "--email", "charlie@example.com")

	var contacts []map[string]interface{}
	tc.runJSON(&contacts, "contact", "list", "--format", "json")
	if len(contacts) != 3 {
		t.Fatalf("expected 3 contacts, got %d", len(contacts))
	}
}

func TestContactList_FilterByTag(t *testing.T) {
	tc := newTestContext(t)

	tc.runOK("contact", "add", "--name", "Alice", "--email", "alice@example.com", "--tag", "vip")
	tc.runOK("contact", "add", "--name", "Bob", "--email", "bob@example.com")

	var contacts []map[string]interface{}
	tc.runJSON(&contacts, "contact", "list", "--tag", "vip", "--format", "json")
	if len(contacts) != 1 {
		t.Fatalf("expected 1 contact with tag vip, got %d", len(contacts))
	}
}

func TestContactList_FilterByCompany(t *testing.T) {
	tc := newTestContext(t)

	tc.runOK("contact", "add", "--name", "Alice", "--email", "alice@acme.com", "--company", "Acme")
	tc.runOK("contact", "add", "--name", "Bob", "--email", "bob@other.com", "--company", "Other")

	var contacts []map[string]interface{}
	tc.runJSON(&contacts, "contact", "list", "--company", "Acme", "--format", "json")
	if len(contacts) != 1 {
		t.Fatalf("expected 1 contact at Acme, got %d", len(contacts))
	}
}

func TestContactList_Sort(t *testing.T) {
	tc := newTestContext(t)

	tc.runOK("contact", "add", "--name", "Charlie")
	tc.runOK("contact", "add", "--name", "Alice")
	tc.runOK("contact", "add", "--name", "Bob")

	var contacts []map[string]interface{}
	tc.runJSON(&contacts, "contact", "list", "--sort", "name", "--format", "json")
	names := make([]string, len(contacts))
	for i, c := range contacts {
		names[i] = c["name"].(string)
	}
	if names[0] != "Alice" || names[1] != "Bob" || names[2] != "Charlie" {
		t.Fatalf("expected alphabetical order, got: %v", names)
	}
}

func TestContactList_LimitOffset(t *testing.T) {
	tc := newTestContext(t)

	tc.runOK("contact", "add", "--name", "A")
	tc.runOK("contact", "add", "--name", "B")
	tc.runOK("contact", "add", "--name", "C")
	tc.runOK("contact", "add", "--name", "D")

	var contacts []map[string]interface{}
	tc.runJSON(&contacts, "contact", "list", "--limit", "2", "--format", "json")
	if len(contacts) != 2 {
		t.Fatalf("expected 2 contacts with limit, got %d", len(contacts))
	}

	tc.runJSON(&contacts, "contact", "list", "--limit", "2", "--offset", "2", "--format", "json")
	if len(contacts) != 2 {
		t.Fatalf("expected 2 contacts with offset, got %d", len(contacts))
	}
}

func TestContactList_FormatIDs(t *testing.T) {
	tc := newTestContext(t)

	tc.runOK("contact", "add", "--name", "Alice")
	tc.runOK("contact", "add", "--name", "Bob")

	out := tc.runOK("contact", "list", "--format", "ids")
	lines := strings.Split(strings.TrimSpace(out), "\n")
	if len(lines) != 2 {
		t.Fatalf("expected 2 IDs, got %d", len(lines))
	}
	for _, line := range lines {
		if !strings.HasPrefix(line, "ct_") {
			t.Errorf("expected ID starting with ct_, got: %s", line)
		}
	}
}

func TestContactList_FormatCSV(t *testing.T) {
	tc := newTestContext(t)

	tc.runOK("contact", "add", "--name", "Alice", "--email", "alice@example.com")

	out := tc.runOK("contact", "list", "--format", "csv")
	lines := strings.Split(strings.TrimSpace(out), "\n")
	if len(lines) < 2 {
		t.Fatalf("expected header + data row, got %d lines", len(lines))
	}
	// First line should be headers.
	assertContains(t, lines[0], "name")
	assertContains(t, lines[0], "email")
	// Second line should have the data.
	assertContains(t, lines[1], "Alice")
	assertContains(t, lines[1], "alice@example.com")
}

func TestContactEdit(t *testing.T) {
	tc := newTestContext(t)

	id := strings.TrimSpace(tc.runOK("contact", "add", "--name", "Jane Doe", "--email", "jane@acme.com", "--title", "Engineer"))

	tc.runOK("contact", "edit", id, "--name", "Jane Smith", "--title", "CTO")

	show := tc.runOK("contact", "show", id)
	assertContains(t, show, "Jane Smith")
	assertContains(t, show, "CTO")
	assertNotContains(t, show, "Jane Doe")
}

func TestContactEdit_ByEmail(t *testing.T) {
	tc := newTestContext(t)

	tc.runOK("contact", "add", "--name", "Jane Doe", "--email", "jane@acme.com")
	tc.runOK("contact", "edit", "jane@acme.com", "--title", "CEO")

	show := tc.runOK("contact", "show", "jane@acme.com")
	assertContains(t, show, "CEO")
}

func TestContactEdit_CustomFields(t *testing.T) {
	tc := newTestContext(t)

	id := strings.TrimSpace(tc.runOK("contact", "add", "--name", "Jane", "--set", "github=janedoe"))

	// Update a custom field.
	tc.runOK("contact", "edit", id, "--set", "github=janesmith")
	show := tc.runOK("contact", "show", id)
	assertContains(t, show, "janesmith")

	// Remove a custom field.
	tc.runOK("contact", "edit", id, "--unset", "github")
	show = tc.runOK("contact", "show", id)
	assertNotContains(t, show, "github")
}

func TestContactEdit_Tags(t *testing.T) {
	tc := newTestContext(t)

	id := strings.TrimSpace(tc.runOK("contact", "add", "--name", "Jane", "--tag", "lead"))

	tc.runOK("contact", "edit", id, "--add-tag", "vip", "--rm-tag", "lead")

	show := tc.runOK("contact", "show", id)
	assertContains(t, show, "vip")
	assertNotContains(t, show, "lead")
}

func TestContactRm(t *testing.T) {
	tc := newTestContext(t)

	id := strings.TrimSpace(tc.runOK("contact", "add", "--name", "Jane", "--email", "jane@acme.com"))

	tc.runOK("contact", "rm", id, "--force")

	_, _ = tc.runFail("contact", "show", id)
}

func TestContactRm_ByEmail(t *testing.T) {
	tc := newTestContext(t)

	tc.runOK("contact", "add", "--name", "Jane", "--email", "jane@acme.com")
	tc.runOK("contact", "rm", "jane@acme.com", "--force")
	_, _ = tc.runFail("contact", "show", "jane@acme.com")
}

func TestContactList_Filter(t *testing.T) {
	tc := newTestContext(t)

	tc.runOK("contact", "add", "--name", "Alice", "--title", "CTO", "--source", "conference")
	tc.runOK("contact", "add", "--name", "Bob", "--title", "Engineer", "--source", "inbound")
	tc.runOK("contact", "add", "--name", "Charlie", "--title", "CTO", "--source", "inbound")

	var contacts []map[string]interface{}
	tc.runJSON(&contacts, "contact", "list", "--filter", "title=CTO AND source=inbound", "--format", "json")
	if len(contacts) != 1 {
		t.Fatalf("expected 1 contact matching filter, got %d", len(contacts))
	}
}

func TestContactMerge(t *testing.T) {
	tc := newTestContext(t)

	id1 := strings.TrimSpace(tc.runOK("contact", "add", "--name", "Jane Doe", "--email", "jane@acme.com", "--tag", "vip"))
	id2 := strings.TrimSpace(tc.runOK("contact", "add", "--name", "J. Doe", "--email", "jane.doe@gmail.com", "--tag", "enterprise"))

	tc.runOK("contact", "merge", id1, id2, "--keep-first")

	// Primary contact should have both emails and both tags.
	show := tc.runOK("contact", "show", id1)
	assertContains(t, show, "jane@acme.com")
	assertContains(t, show, "jane.doe@gmail.com")
	assertContains(t, show, "vip")
	assertContains(t, show, "enterprise")

	// Second contact should no longer exist.
	_, _ = tc.runFail("contact", "show", id2)
}

File: test/contact.test.ts
import { describe, expect, test } from 'bun:test'

import { createTestContext } from './helpers.ts'

describe('contact add', () => {
  test('basic add returns prefixed ID', () => {
    const ctx = createTestContext()
    const out = ctx.runOK('contact', 'add', '--name', 'Jane Doe')
    expect(out.trim()).toStartWith('ct_')
  })

  test('full add stores all fields', () => {
    const ctx = createTestContext()
    const id = ctx
      .runOK(
        'contact', 'add',
        '--name', 'Jane Doe',
        '--email', 'jane@acme.com',
        '--phone', '+1-555-0100',
        '--company', 'Acme Corp',
        '--title', 'CTO',
        '--source', 'conference',
        '--tag', 'hot-lead',
        '--tag', 'enterprise',
        '--set', 'linkedin=linkedin.com/in/janedoe',
      )
      .trim()

    const show = ctx.runOK('contact', 'show', id)
    expect(show).toContain('Jane Doe')
    expect(show).toContain('jane@acme.com')
    expect(show).toContain('+1-555-0100')
    expect(show).toContain('CTO')
    expect(show).toContain('conference')
    expect(show).toContain('hot-lead')
    expect(show).toContain('enterprise')
    expect(show).toContain('linkedin.com/in/janedoe')
  })

  test('fails without --name', () => {
    const ctx = createTestContext()
    const result = ctx.runFail('contact', 'add', '--email', 'nobody@example.com')
    expect(result.stderr).toContain('name')
  })

  test('rejects duplicate email', () => {
    const ctx = createTestContext()
    ctx.runOK('contact', 'add', '--name', 'Jane Doe', '--email', 'jane@acme.com')
    const result = ctx.runFail('contact', 'add', '--name', 'Jane Smith', '--email', 'jane@acme.com')
    expect(result.stderr).toContain('duplicate')
  })

  test('multiple emails on create', () => {
    const ctx = createTestContext()
    const id = ctx.runOK(
      'contact', 'add', '--name', 'Jane Doe',
      '--email', 'jane@acme.com', '--email', 'jane.doe@gmail.com',
    ).trim()

    const show = ctx.runOK('contact', 'show', id)
    expect(show).toContain('jane@acme.com')
    expect(show).toContain('jane.doe@gmail.com')
  })

  test('multiple phones on create', () => {
    const ctx = createTestContext()
    const id = ctx.runOK(
      'contact', 'add', '--name', 'Jane Doe',
      '--phone', '+1-555-0100', '--phone', '+44-20-7946-0958',
    ).trim()

    const show = ctx.runOK('contact', 'show', id)
    expect(show).toContain('+1-555-0100')
    expect(show).toContain('+44-20-7946-0958')
  })

  test('lookup by any email when contact has multiple', () => {
    const ctx = createTestContext()
    ctx.runOK(
      'contact', 'add', '--name', 'Jane Doe',
      '--email', 'jane@acme.com', '--email', 'jane.doe@gmail.com',
    )

    const show1 = ctx.runOK('contact', 'show', 'jane@acme.com')
    const show2 = ctx.runOK('contact', 'show', 'jane.doe@gmail.com')
    expect(show1).toContain('Jane Doe')
    expect(show2).toContain('Jane Doe')
  })

  test('duplicate check applies across all emails', () => {
    const ctx = createTestContext()
    ctx.runOK('contact', 'add', '--name', 'Jane', '--email', 'jane@acme.com', '--email', 'jane@personal.com')
    // Adding a new contact with jane@personal.com should fail — it belongs to Jane.
    const result = ctx.runFail('contact', 'add', '--name', 'Other Jane', '--email', 'jane@personal.com')
    expect(result.stderr).toContain('duplicate')
  })
})

describe('contact show', () => {
  test('by email', () => {
    const ctx = createTestContext()
    ctx.runOK('contact', 'add', '--name', 'Jane Doe', '--email', 'jane@acme.com')
    const out = ctx.runOK('contact', 'show', 'jane@acme.com')
    expect(out).toContain('Jane Doe')
  })

  test('not found returns error', () => {
    const ctx = createTestContext()
    ctx.runFail('contact', 'show', 'nonexistent@example.com')
  })
})

describe('contact list', () => {
  test('empty database returns empty array', () => {
    const ctx = createTestContext()
    const contacts = ctx.runJSON<unknown[]>('contact', 'list', '--format', 'json')
    expect(contacts).toEqual([])
  })

  test('returns all contacts', () => {
    const ctx = createTestContext()
    ctx.runOK('contact', 'add', '--name', 'Alice', '--email', 'alice@example.com')
    ctx.runOK('contact', 'add', '--name', 'Bob', '--email', 'bob@example.com')
    ctx.runOK('contact', 'add', '--name', 'Charlie', '--email', 'charlie@example.com')

    const contacts = ctx.runJSON<unknown[]>('contact', 'list', '--format', 'json')
    expect(contacts).toHaveLength(3)
  })

  test('filter by tag', () => {
    const ctx = createTestContext()
    ctx.runOK('contact', 'add', '--name', 'Alice', '--email', 'alice@example.com', '--tag', 'vip')
    ctx.runOK('contact', 'add', '--name', 'Bob', '--email', 'bob@example.com')

    const contacts = ctx.runJSON<unknown[]>('contact', 'list', '--tag', 'vip', '--format', 'json')
    expect(contacts).toHaveLength(1)
  })

  test('filter by company', () => {
    const ctx = createTestContext()
    ctx.runOK('contact', 'add', '--name', 'Alice', '--email', 'alice@acme.com', '--company', 'Acme')
    ctx.runOK('contact', 'add', '--name', 'Bob', '--email', 'bob@other.com', '--company', 'Other')

    const contacts = ctx.runJSON<unknown[]>('contact', 'list', '--company', 'Acme', '--format', 'json')
    expect(contacts).toHaveLength(1)
  })

  test('sort by name', () => {
    const ctx = createTestContext()
    ctx.runOK('contact', 'add', '--name', 'Charlie')
    ctx.runOK('contact', 'add', '--name', 'Alice')
    ctx.runOK('contact', 'add', '--name', 'Bob')

    const contacts = ctx.runJSON<Array<{ name: string }>>('contact', 'list', '--sort', 'name', '--format', 'json')
    expect(contacts.map((c) => c.name)).toEqual(['Alice', 'Bob', 'Charlie'])
  })

  test('limit and offset', () => {
    const ctx = createTestContext()
    ctx.runOK('contact', 'add', '--name', 'A')
    ctx.runOK('contact', 'add', '--name', 'B')
    ctx.runOK('contact', 'add', '--name', 'C')
    ctx.runOK('contact', 'add', '--name', 'D')

    const page1 = ctx.runJSON<unknown[]>('contact', 'list', '--limit', '2', '--format', 'json')
    expect(page1).toHaveLength(2)

    const page2 = ctx.runJSON<unknown[]>('contact', 'list', '--limit', '2', '--offset', '2', '--format', 'json')
    expect(page2).toHaveLength(2)
  })

  test('format ids outputs one ID per line', () => {
    const ctx = createTestContext()
    ctx.runOK('contact', 'add', '--name', 'Alice')
    ctx.runOK('contact', 'add', '--name', 'Bob')

    const out = ctx.runOK('contact', 'list', '--format', 'ids')
    const lines = out.trim().split('\n')
    expect(lines).toHaveLength(2)
    for (const line of lines) {
      expect(line).toStartWith('ct_')
    }
  })

  test('format csv has header and data rows', () => {
    const ctx = createTestContext()
    ctx.runOK('contact', 'add', '--name', 'Alice', '--email', 'alice@example.com')

    const out = ctx.runOK('contact', 'list', '--format', 'csv')
    const lines = out.trim().split('\n')
    expect(lines.length).toBeGreaterThanOrEqual(2)
    expect(lines[0]).toContain('name')
    expect(lines[0]).toContain('email')
    expect(lines[1]).toContain('Alice')
    expect(lines[1]).toContain('alice@example.com')
  })

  test('filter expression', () => {
    const ctx = createTestContext()
    ctx.runOK('contact', 'add', '--name', 'Alice', '--title', 'CTO', '--source', 'conference')
    ctx.runOK('contact', 'add', '--name', 'Bob', '--title', 'Engineer', '--source', 'inbound')
    ctx.runOK('contact', 'add', '--name', 'Charlie', '--title', 'CTO', '--source', 'inbound')

    const contacts = ctx.runJSON<unknown[]>(
      'contact', 'list', '--filter', 'title=CTO AND source=inbound', '--format', 'json',
    )
    expect(contacts).toHaveLength(1)
  })
})

describe('contact edit', () => {
  test('update fields by ID', () => {
    const ctx = createTestContext()
    const id = ctx.runOK('contact', 'add', '--name', 'Jane Doe', '--email', 'jane@acme.com', '--title', 'Engineer').trim()
    ctx.runOK('contact', 'edit', id, '--name', 'Jane Smith', '--title', 'CTO')

    const show = ctx.runOK('contact', 'show', id)
    expect(show).toContain('Jane Smith')
    expect(show).toContain('CTO')
    expect(show).not.toContain('Jane Doe')
  })

  test('update by email', () => {
    const ctx = createTestContext()
    ctx.runOK('contact', 'add', '--name', 'Jane Doe', '--email', 'jane@acme.com')
    ctx.runOK('contact', 'edit', 'jane@acme.com', '--title', 'CEO')

    const show = ctx.runOK('contact', 'show', 'jane@acme.com')
    expect(show).toContain('CEO')
  })

  test('set and unset custom fields', () => {
    const ctx = createTestContext()
    const id = ctx.runOK('contact', 'add', '--name', 'Jane', '--set', 'github=janedoe').trim()

    ctx.runOK('contact', 'edit', id, '--set', 'github=janesmith')
    expect(ctx.runOK('contact', 'show', id)).toContain('janesmith')

    ctx.runOK('contact', 'edit', id, '--unset', 'github')
    expect(ctx.runOK('contact', 'show', id)).not.toContain('github')
  })

  test('add and remove tags', () => {
    const ctx = createTestContext()
    const id = ctx.runOK('contact', 'add', '--name', 'Jane', '--tag', 'lead').trim()
    ctx.runOK('contact', 'edit', id, '--add-tag', 'vip', '--rm-tag', 'lead')

    const show = ctx.runOK('contact', 'show', id)
    expect(show).toContain('vip')
    expect(show).not.toContain('lead')
  })

  test('add email to existing contact', () => {
    const ctx = createTestContext()
    const id = ctx.runOK('contact', 'add', '--name', 'Jane', '--email', 'jane@acme.com').trim()
    ctx.runOK('contact', 'edit', id, '--add-email', 'jane.doe@gmail.com')

    const show = ctx.runOK('contact', 'show', id)
    expect(show).toContain('jane@acme.com')
    expect(show).toContain('jane.doe@gmail.com')
  })

  test('remove email from contact', () => {
    const ctx = createTestContext()
    const id = ctx.runOK('contact', 'add', '--name', 'Jane', '--email', 'jane@acme.com', '--email', 'old@acme.com').trim()
    ctx.runOK('contact', 'edit', id, '--rm-email', 'old@acme.com')

    const show = ctx.runOK('contact', 'show', id)
    expect(show).toContain('jane@acme.com')
    expect(show).not.toContain('old@acme.com')
  })

  test('add phone to existing contact', () => {
    const ctx = createTestContext()
    const id = ctx.runOK('contact', 'add', '--name', 'Jane', '--phone', '+1-555-0100').trim()
    ctx.runOK('contact', 'edit', id, '--add-phone', '+44-20-7946-0958')

    const show = ctx.runOK('contact', 'show', id)
    expect(show).toContain('+1-555-0100')
    expect(show).toContain('+44-20-7946-0958')
  })

  test('remove phone from contact', () => {
    const ctx = createTestContext()
    const id = ctx.runOK('contact', 'add', '--name', 'Jane', '--phone', '+1-555-0100', '--phone', '+1-555-OLD').trim()
    ctx.runOK('contact', 'edit', id, '--rm-phone', '+1-555-OLD')

    const show = ctx.runOK('contact', 'show', id)
    expect(show).toContain('+1-555-0100')
    expect(show).not.toContain('+1-555-OLD')
  })
})

describe('contact rm', () => {
  test('delete by ID', () => {
    const ctx = createTestContext()
    const id = ctx.runOK('contact', 'add', '--name', 'Jane', '--email', 'jane@acme.com').trim()
    ctx.runOK('contact', 'rm', id, '--force')
    ctx.runFail('contact', 'show', id)
  })

  test('delete by email', () => {
    const ctx = createTestContext()
    ctx.runOK('contact', 'add', '--name', 'Jane', '--email', 'jane@acme.com')
    ctx.runOK('contact', 'rm', 'jane@acme.com', '--force')
    ctx.runFail('contact', 'show', 'jane@acme.com')
  })
})

describe('contact merge', () => {
  test('merges two contacts keeping first', () => {
    const ctx = createTestContext()
    const id1 = ctx.runOK('contact', 'add', '--name', 'Jane Doe', '--email', 'jane@acme.com', '--tag', 'vip').trim()
    const id2 = ctx.runOK('contact', 'add', '--name', 'J. Doe', '--email', 'jane.doe@gmail.com', '--tag', 'enterprise').trim()

    ctx.runOK('contact', 'merge', id1, id2, '--keep-first')

    const show = ctx.runOK('contact', 'show', id1)
    expect(show).toContain('jane@acme.com')
    expect(show).toContain('jane.doe@gmail.com')
    expect(show).toContain('vip')
    expect(show).toContain('enterprise')

    ctx.runFail('contact', 'show', id2)
  })
})
